site/components/pages/Pledge/lib/verifyGnosisSafeMultisig.ts

```diff
@@ -0,0 +1,60 @@
+import { ethers } from "ethers";
+import { getJsonRpcProvider } from "@klimadao/lib/utils";
+import GnosisSafeSignMessageLib from "@klimadao/lib/abi/GnosisSafeSignMessageLib.json";
+
+const ONE_HOUR = 60 * 60 * 1000;
+
+export const waitForGnosisSignature = async (params: {
+  message: string;
+  address: string;
+}): Promise<void> => {
+  const provider = getJsonRpcProvider();
+  const gnosisSafeContract = new ethers.Contract(
+    params.address,
+    GnosisSafeSignMessageLib.abi,
+    provider
+  );
+  const messageHash = ethers.utils.hashMessage(params.message);
+  const getMessageHash = await gnosisSafeContract.getMessageHash(messageHash);
+  const signedEvent = gnosisSafeContract.filters.SignMsg(getMessageHash);
+
+  const waitForSignedEvent = (): Promise<void> =>
+    new Promise((resolve, reject) => {
+      setTimeout(() => {
+        gnosisSafeContract.removeListener(signedEvent, () => resolve());
+        reject(new Error("Gnosis safe signature verification timed out"));
+      }, ONE_HOUR);
+
+      gnosisSafeContract.once(signedEvent, () => resolve());
+    });
+
+  await waitForSignedEvent();
+};
+
+export const verifyGnosisSignature = async (params: {
+  /** Plain un-hashed string with expected nonce */
+  message: string;
+  /** Address of the multisig contract */
+  address: string;
+}) => {
+  const provider = getJsonRpcProvider();
+  const gnosisSafeContract = new ethers.Contract(
+    params.address,
+    GnosisSafeSignMessageLib.abi,
+    provider
+  );
+  const messageHash = ethers.utils.hashMessage(params.message);
+  const getMessageHash = await gnosisSafeContract.getMessageHash(messageHash);
```

Comment:
RPC can be slow and flaky, so we should calculate this inline using ethers rather than calling the contract

Reply:
Ah gotcha, so that's why code was pulled in from the other gnosis' repo to leverage ethers.
Makes sense :)